### ISSUE TITLE
Max active streams reached exception

### ISSUE THREAD

**Issue by nicsam1997** [reactor/reactor-netty#2981]
## Expected Behavior
When a connection observes an exception it should stopped being used

## Actual Behavior
The bad connection remains open for a long time until it is eventually closed

## Steps to Reproduce
Cannot reproduce, have only seen it in production. Have attached logs below. 
[logs_reactor_netty_remove_connection.txt](https://github.com/reactor/reactor-netty/files/13440490/logs_reactor_netty_remove_connection.txt)
[logs_reactor_netty_start_of_issue.txt](https://github.com/reactor/reactor-netty/files/13440492/logs_reactor_netty_start_of_issue.txt)
[logs_reactor_netty_streams_increasing.txt](https://github.com/reactor/reactor-netty/files/13440494/logs_reactor_netty_streams_increasing.txt)

## Your Environment

* Reactor version(s) used: spring-boot-starter-parent:3.1.5 spring-boot-start-reactor-netty:3.1.5, reactor-netty-http:1.1.12
* JVM version (`java -version`): 17
* OS and version (eg. `uname -a`):

**Comment by violetagg** [comment]
@nicsam1997 Are there other logs from `reactor.netty.http.client.Http2Pool` logger?

**Comment by violetagg** [comment]
@nicsam1997 In Gitter thread you mentioned `ReadTimeoutExceptions ` . Can you add also this logs and the configuration that you have in order to receive `ReadTimeoutExceptions`?
https://matrix.to/#/!rdZMIWMDXqzVEagqCt:gitter.im/$wavvZKbYc5hXnehfGjiExvduwSMa746rQH8ONG2QBWQ?via=gitter.im&via=matrix.org&via=mozilla.org

**Comment by nicsam1997** [comment]
The configuration is like this: We have a responseTimeout of 4s
```
final var httpClient =
    HttpClient.create(providerBuilder.build())
        .protocol(config.http2Enabled() ? HttpProtocol.H2 : HttpProtocol.HTTP11)
        .responseTimeout(config.responseTimeout())
        .option(
            CONNECT_TIMEOUT_MILLIS,
            Long.valueOf(config.connectTimeout().toMillis()).intValue());
httpClient.warmup().block();
final var contextSpec =
  Http2SslContextSpec.forClient()
    .configure(
      builder ->
        configureKeyManager(keyStoreFullPath, mtlsConfig.keyStorePassword(), builder));
return httpClient.secure(t -> t.sslContext(contextSpec));
```

**Comment by nicsam1997** [comment]
There are logs for the Http2Pool logger towards the end of the issue, I have filtered out the logs containing "Channel deactivated" or "Channel activated"
[logs_rector_netty_include_errors.txt](https://github.com/reactor/reactor-netty/files/13449532/logs_rector_netty_include_errors.txt)

**Comment by nicsam1997** [comment]
Hi @violetagg Have now debugged this issue a bit further and managed to reproduce. Basically it seems like the ReadTimeOutException is thrown but this does not lead to the connection being closed. I am not sure if this is intended or not but would want to customize the behaviour, basically as soon as I observe a ReadTimeOutException I want to dispose the used connection. Is this possible?

**Comment by violetagg** [comment]
@nicsam1997 Are you able to share this repro?

**Comment by nicsam1997** [comment]
Yes, but probably tomorrow, need to clean it up a bit

**Comment by nicsam1997** [comment]
@violetagg Edited above

**Comment by nicsam1997** [comment]
Hi, have created a reproduction here: https://github.com/nicsam1997/http2reproduce

**Comment by violetagg** [comment]
@nicsam1997 IMO what you observe with this reproducible example is a bit different. Let me explain: based only on response timeout, we cannot consider that the connection itself is broken (HTTP/2 use case -> `ReadTimeoutException` is received on the level of stream). Reactor Netty gives you the ability to specify response timeout per request, which means different requests might have different timeouts (of course you may also specify one global/default response timeout that will apply to all requests, which is the case here). Some calls might reach their timeout if they request some slow resources on the server, but other might not reach their timeouts if the server is able to respond within the timeout.

Your use case is more in the group of `network component drops silently the connection/packets etc.` (firewall/loadbalancer etc.). For these kind of issues you can enable `SO_KEEPALIVE`, more about this you can find in the reference documentation:
https://projectreactor.io/docs/netty/release/reference/index.html#faq.connection-closed
https://projectreactor.io/docs/netty/release/reference/index.html#connection-timeout

I added this configuration to your example:

```
        final var httpClient =
                HttpClient.create(connectionProvider)
                        .protocol(HttpProtocol.H2C)
                        .responseTimeout(Duration.ofSeconds(4))
                        // SO_KEEPALIVE configuration start
                        .option(ChannelOption.SO_KEEPALIVE, true)
                        .option(NioChannelOption.of(ExtendedSocketOptions.TCP_KEEPIDLE), 1)
                        .option(NioChannelOption.of(ExtendedSocketOptions.TCP_KEEPINTERVAL), 1)
                        .option(NioChannelOption.of(ExtendedSocketOptions.TCP_KEEPCOUNT), 8)
                        // SO_KEEPALIVE configuration end
                        .option(CONNECT_TIMEOUT_MILLIS, 50)
                        .option(EpollChannelOption.TCP_USER_TIMEOUT, 3000);
```

**Comment by nicsam1997** [comment]
Keep-alive does not help me get rid of a busy connection, right? Ideally the connection would not break but if it does, would this setting do anything to remove it from the pool (given that I already have traffic on the connection?)

**Comment by violetagg** [comment]
As I wrote I think the reproducible example is a bit different - there `SO_KEEPALIVE` helps because you have firewall dropping packages.

**Comment by violetagg** [comment]
@nicsam1997 Based on the provided logs and a code review, I found one issue (fixed with #3031). Will you be able to test the current snapshots? (1.0.42-SNAPSHOT/1.1.16-SNAPSHOT)

**Comment by nicsam1997** [comment]
Possibly, but what does this bugfix fix? I have not seen the max active streams exception for a long time now so that will be hard to test

**Comment by violetagg** [comment]
It tries to fix the reported problem?
 
`Immediately aborted pooled channel, max active streams is reached, re-acquiring a new channel`
`Error while acquiring from reactor.netty.http.client.Http2Pool@41130dcb. Max active streams is reached.`

**Comment by nicsam1997** [comment]
Since I don’t have any other way to reproduce but by running in production and wait for a long time I think testing the snapshot will be a bit hard

**Comment by violetagg** [comment]
@nicsam1997 I'm going to close this issue as I think that #3031 should fix it. The fix will be available in 1.0.42/1.1.16 (the planned date for these releases is 13.02). When the release is available and you are able to upgrade, please test it. If the problem is still there we can reopen this issue.
